packages-exp/auth-exp/src/platform_browser/popup_redirect.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  AuthEventType,
+  EventManager,
+  GapiAuthEvent,
+  GapiOutcome,
+  PopupRedirectResolver
+} from '../model/popup_redirect';
+import { _openIframe } from './iframe/iframe';
+
+/**
+ * URL for Authentication widget which will initiate the OAuth handshake
+ */
+const WIDGET_URL = '__/auth/handler';
+
+export class BrowserPopupRedirectResolver implements PopupRedirectResolver {
+  private eventManager: EventManager | null = null;
+
+  // Wrapping in async even though we don't await anywhere in order
+  // to make sure errors are raised as promise rejections
+  async openPopup(
+    auth: Auth,
+    provider: externs.AuthProvider,
+    authType: AuthEventType,
+    eventId?: string
+  ): Promise<AuthPopup> {
+    const url = getRedirectUrl(auth, provider, authType, eventId);
+    return _open(auth.name, url, _generateEventId());
+  }
+
+  async initialize(auth: Auth): Promise<EventManager> {
+    if (this.eventManager) {
+      return this.eventManager;
+    }
+
+    const iframe = await _openIframe(auth);
+    const eventManager = new AuthEventManager();
+    iframe.register<GapiAuthEvent>(
+      'authEvent',
+      async (message: GapiAuthEvent) => {
+        await eventManager.onEvent(message.authEvent);
+
+        // We always ACK with the iframe
+        return { status: GapiOutcome.ACK };
+      },
+      gapi.iframes.CROSS_ORIGIN_IFRAMES_FILTER
+    );
+
+    this.eventManager = eventManager;
+    return eventManager;
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+type WidgetParams = {
+  apiKey: ApiKey;
+  appName: AppName;
+  authType: AuthEventType;
+  redirectUrl: string;
+  v: string;
+  providerId?: string;
+  scopes?: string;
+  customParameters?: string;
+  eventId?: string;
+};
+
+function getRedirectUrl(
+  auth: Auth,
+  provider: externs.AuthProvider,
+  authType: AuthEventType,
+  eventId?: string
+): string {
+  assert(auth.config.authDomain, auth.name, AuthErrorCode.MISSING_AUTH_DOMAIN);
+  assert(auth.config.apiKey, auth.name, AuthErrorCode.INVALID_API_KEY);
+
+  const params: WidgetParams = {
+    apiKey: auth.config.apiKey,
+    appName: auth.name,
+    authType,
+    redirectUrl: _getCurrentUrl(),
+    v: SDK_VERSION,
+    eventId
+  };
+
+  if (provider instanceof OAuthProvider) {
+    provider.setDefaultLanguage(auth.languageCode);
+    params.providerId = provider.providerId || '';
+    if (!isEmpty(provider.getCustomParameters())) {
+      params.customParameters = JSON.stringify(provider.getCustomParameters());
+    }
+    const scopes = provider.getScopes();
+    if (scopes.length > 0) {
+      params.scopes = scopes.join(',');
+    }
+    // TODO set additionalParams?
+    // let additionalParams = provider.getAdditionalParams();
+    // for (let key in additionalParams) {
+    //   if (!params.hasOwnProperty(key)) {
+    //     params[key] = additionalParams[key]
+    //   }
+    // }
+  }
+
+  // TODO: maybe set tid as tenantId
```

Comment:
we should definitely do this at least now that we have it on auth

Reply:
Done